### ISSUE TITLE
allOf Inheritance is broken in schema with more than two layers hierarchy

### ISSUE THREAD

**Issue by globalstick** [OpenAPITools/openapi-generator#4423]
#### Bug Report Checklist

- [ ] Have you provided a full/minimal spec to reproduce the issue?
- [ ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] What's the version of OpenAPI Generator used?
- [ ] Have you search for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

In schema hierarchy with more, than two layers by Code Generation all classes in parent classes tree are merged within child class.

For example, if you try to generate a model classes from SchemaC, which allOfs SchemaB  (with discrimanator), and SchemaB (with discrimanator) allofs SchemaA  (with discrimanator), you will get Java class SchemaC, which contains all attributes of SchemaC, SchemaB and SchemaA instead of extending class SchemaB. SchemaC does not extend any class at all. Could you please solve this issue. Some example code is down there (without root yaml file)

##### openapi-generator version

4.4.2

##### OpenAPI declaration file content or url

<!-- if it is a bug, a json or yaml that produces it.
If you post the code inline, please wrap it with
````yaml
Creature:
   type: object
   required:
      - id
      - kind
   properties:
      id:
         description: Identificator
         type: string
      kind:
         description: Kind of creature
         type: string
      discriminator:
         propertyName: kind

Human:
     allOf:
        - $ref: '#/components/schemas/Creature'
        - type: object
        required:
           - name
           - surname
        properties:
           name:
              type: string
           surname:
               type: string
           address:
               type: string
        discriminator:
           propertyName: surname

Student:
   allOf:
      - $ref: '#/components/schemas/Human'
      - type: object
        required:
           - college
           - subject
        properties:
           college:
              type: string
           subject:
              type: string  
  
```

##### Command line used for generation

Maven Plugin generator

##### Steps to reproduce

Create three level schema inheritance hierarchy (with discriminator) and generate Java code 

##### Related issues/PRs

##### Suggest a fix
Each class has only ist own Attributes and extend the allOf schema class.

**Comment by roxspring** [comment]
After attempting a migration from swagger-codegen 2.x I hit this yesterday morning and have been scratching my head ever since. I concluded that its a bug worth reporting and discovered that one was raised just 2h earlier!

Applying the problem to the petstore examples, I [created a `BigCat` definition extending `Cat` which extends `Animal`](https://github.com/roxspring/openapi-generator/commit/394dc66176a3c93b4d18a38ccdd96a05cc36e21c) (without bothering with an additional discriminator, but valid according to https://apitools.dev/swagger-parser/online/), and ran `./bin/springboot-petstore-server.sh`:

```
...
  BigCat:
    allOf:
      - $ref: '#/definitions/Cat'
      - type: object
        properties:
          kind:
            type: string
            enum: [lions, tigers, leopards, jaguars]
...
```

Under swagger-codegen 2.3.1, 2.4.9, `Animal` is annotated:
```
@JsonSubTypes({
  @JsonSubTypes.Type(value = BigCat.class, name = "BigCat"),
  @JsonSubTypes.Type(value = Dog.class, name = "Dog"),
  @JsonSubTypes.Type(value = Cat.class, name = "Cat"),
})
```

and `BigCat` is defined:

```
public class BigCat extends Cat  {
  /**
   * Gets or Sets kind
   */
  public enum KindEnum {
    LIONS("lions"),    
    TIGERS("tigers"),    
    LEOPARDS("leopards"),    
    JAGUARS("jaguars");
...
```

Whereas under openapi-generator `Animal` annotations know nothing of `BigCat`:

```
@JsonSubTypes({
  @JsonSubTypes.Type(value = Dog.class, name = "Dog"),
  @JsonSubTypes.Type(value = Cat.class, name = "Cat"),
})
```

and `BigCat` is defined without reference to `Cat` and copying all the fields instead:

```
public class BigCat   {
  @JsonProperty("className")
  private String className;

  @JsonProperty("color")
  private String color = "red";

  @JsonProperty("declawed")
  private Boolean declawed;
...
```

Not sure whether this ever worked in openapi-generator but it's certainly a regression since forking.

Also odd: openapi-generator generates additional `DogAllOf`, `CatAllOf`, `BigCatAllOf` seemingly coming from the composite schemas. Maybe this is unrelated and the new normal but it's really weird.

**Comment by roxspring** [comment]
... and researching further it appears this might be the same as previously reported variants:

- https://github.com/OpenAPITools/openapi-generator/issues/2928
- https://github.com/OpenAPITools/openapi-generator/issues/3058
- https://github.com/OpenAPITools/openapi-generator/issues/1685

None of these seem to have any activity or even acknowledgement 😞. Is there any chance of this regression being addressed @wing328? Or can somebody offer guidance or some starting points on what the fix might be?

**Comment by cvgaviao** [comment]
@wing328, could you please take a look on this issue and the provided PR?

**Comment by jimschubert** [comment]
Fixed by #4503 please evaluate and let us know if it needs to be reopened.
